serverjs/cubefn.js

```diff
@@ -228,6 +252,172 @@ async function getElo(cardnames, round) {
+      set: split[4].toUpperCase(),
+      addedTmsp: new Date(),
+      collector_number: split[5],
+      status: split[6],
+      finish: split[7],
+      imgUrl: split[9] && split[9] !== 'undefined' ? split[9] : null,
+      tags: split[10] && split[10].length > 0 ? split[10].split(',') : [],
+    };
+
+    const potentialIds = carddb.allIds(card);
+    if (potentialIds && potentialIds.length > 0) {
+      // First, try to find the correct set.
+      const matchingSetAndNumber = potentialIds.find((id) => {
+        const dbCard = carddb.cardFromId(id);
+        return (
+          card.set.toUpperCase() === dbCard.set.toUpperCase() &&
+          card.collector_number.toUpperCase() === dbCard.collector_number.toUpperCase()
+        );
+      });
+      const matchingSet = potentialIds.find((id) => carddb.cardFromId(id).set.toUpperCase() === card.set);
+      const nonPromo = potentialIds.find(carddb.notPromoOrDigitalId);
+      const first = potentialIds[0];
+      card.cardID = matchingSetAndNumber || matchingSet || nonPromo || first;
+      if (maybeboard === 'true') {
+        newMaybe.push(card);
+      } else {
+        newCards.push(card);
+      }
+    } else {
+      missing += `${card.name}\n`;
+    }
+  }
+  return { newCards, newMaybe, missing };
+}
+
+async function generateBlogpost(req, res, cube, changelog, added, missing, carddb) {
+  const blogpost = new Blog();
+  blogpost.title = 'Cube Bulk Import - Automatic Post';
+  blogpost.html = changelog;
+  blogpost.owner = cube.owner;
+  blogpost.date = Date.now();
+  blogpost.cube = cube._id;
+  blogpost.dev = 'false';
+  blogpost.date_formatted = blogpost.date.toLocaleString('en-US');
+  blogpost.username = cube.owner_name;
+  blogpost.cubename = cube.name;
+
+  if (missing.length > 0) {
+    const reactProps = {
+      cubeID: req.params.id,
+      missing,
+      added: added.map(({ _id, name, image_normal, image_flip }) => ({ _id, name, image_normal, image_flip })),
+      blogpost: blogpost.toObject(),
+    };
+    return res.render('cube/bulk_upload', {
+      reactProps: serialize(reactProps),
+      cube,
+      cube_id: req.params.id,
+      title: `${abbreviate(cube.name)} - Bulk Upload`,
+    });
+  }
+  await blogpost.save();
+  cube = setCubeType(cube, carddb);
+  try {
+    await Cube.updateOne({ _id: cube._id }, cube);
+  } catch (err) {
+    if (err) {
+      req.flash('danger', 'Error adding cards. Please try again.');
+      return res.redirect(`/cube/list/${req.params.id}`);
+    }
+  }
+  req.flash('success', 'All cards successfully added.');
+  return res.redirect(`/cube/list/${req.params.id}`);
+}
+
+// prices should be the prices module with the GetPrices function.
+// elo should be in the form { round: bool }.
+// requested details is a string to pass to carddb.cardFromId.
+async function populateCardDetails(cardLists, carddb, { GetPrices = null, elo = null, requested_details = undefined }) {
```

Comment:
```suggestion
async function populateCardDetails(cardLists, carddb, { prices = false, elo = false, requestedDetails }) {
```
Why does this take a list of lists instead of just a list?

I'm also still not super happy with this giant function - I don't like things with side effects like this. I'd rather err on the side of having some slight repeated code elsewhere in our codebase. Instead, you could imagine a pattern where we have functions like getElo that return `id: { elo }` map and getPrices2 that returns `id: { price, price_foil }` map so that you can easily spread them into an existing card object. That architecture makes more sense to me.

Reply:
It takes a list of lists for the cases where we're working with multiple lists like comparisons. Doing it all in one batch lets us batch things like the price api calls. I did try to prevent this from mutating any of the arguments, and return a new list, but unless we want to do a deep copy it'll have to modify the card objects.